lib/compat/wordpress-6.2/class-gutenberg-rest-block-patterns-controller-6-2.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * REST API: Gutenberg_REST_Block_Patterns_Controller_6_2 class
+ *
+ * @package    Gutenberg
+ * @subpackage REST_API
+ */
+
+/**
+ * Core class used to access block patterns via the REST API.
+ *
+ * @since 6.0.0
+ *
+ * @see WP_REST_Controller
+ */
+class Gutenberg_REST_Block_Patterns_Controller_6_2 extends Gutenberg_REST_Block_Patterns_Controller_6_1 {
+	/**
+	 * An array that maps old categories names to new ones.
+	 *
+	 * @since 6.2.0
+	 * @var array
+	 */
+	protected static $categories_migration = array(
+		'buttons' => 'call-to-action',
+		'query'   => 'posts',
+	);
+
+	/**
+	 * Registers the routes for the objects of the controller.
+	 *
+	 * @since 6.0.0
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			),
+			true
+		);
+	}
+	/**
+	 * Retrieves all block patterns.
+	 *
+	 * @since 6.0.0
+	 * @since 6.2.0 Added migration for old pattern categories.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		if ( ! $this->remote_patterns_loaded ) {
+			// Load block patterns from w.org.
+			_load_remote_block_patterns(); // Patterns with the `core` keyword.
+			_load_remote_featured_patterns(); // Patterns in the `featured` category.
+			_register_remote_theme_patterns(); // Patterns requested by current theme.
+
+			$this->remote_patterns_loaded = true;
+		}
+
+		$response = array();
+		$patterns = WP_Block_Patterns_Registry::get_instance()->get_all_registered();
+		foreach ( $patterns as $pattern ) {
+			$migrated_pattern = $this->migrate_pattern_categories( $pattern );
```

Comment:
Do you think there would be a better place to migrate? Also should we add when a timeline for when to remove this check? Maybe after some WP releases? --cc @jsnajdr @youknowriad

Reply:
> Do you think there would be a better place to migrate?

I think here makes sense for now.

> Also should we add when a timeline for when to remove this check?

We are not in a rush to remove the check, we could keep it for years. So, instead, let's document this particular migration with a `@since` note.

Reply:
I had a `@since` note in the function's docs and added a couple more lines in [this commit](https://github.com/WordPress/gutenberg/pull/46144/commits/4bbacdee9c97ae439a2ea1814a6a6fa5766c2a61). Did you have something more in mind or a rewording?

Reply:
I'd implement this inside the `Gutenberg_REST_Block_Patterns_Controller_6_2::prepare_item_for_response()` method to not add a new method: `Gutenberg_REST_Block_Patterns_Controller_6_2 ::migrate_pattern_categories()`.
But this is just my personal preference.

Also, it doesn't make sense to me to make `Gutenberg_REST_Block_Patterns_Controller_6_2::$categories_migration` and `Gutenberg_REST_Block_Patterns_Controller_6_2::migrate_pattern_categories()` **_protected_** unless there is a plan to extend that class somewhere.
Sorry for not being able to review the PR sooner (I understand it's already merged).

Reply:
I made them protected because they will possibly be updated down the road with the massive revamp of pattern categories and we had similar problems with other class in the past(like in the `theme-json` classes). In those cases we ended up needing core patches to make the protected, because in GB we extend either core classes or per compat version classes in the form of `class_6_2 extends class_6_1`.

>I'd implement this inside the Gutenberg_REST_Block_Patterns_Controller_6_2::prepare_item_for_response()

I could do a follow up for this if you like. The new function is internally called so I wouldn't expect any problems if we remove it even in the future.

Reply:
Thanks for the reply, @ntsekouras.

> I could do a follow up for this if you like.

It's my personal preference and I don't think that's very important.
So maybe let's leave it as is.
Thanks for your willingness to refactor it.